### PR TITLE
feat: static color coding for key column

### LINKED ISSUE #7810: Color code key column in library
Reported by: **[naught101](https://github.com/naught101)**
Date: 2015-01-16T01:35:05Z
Status: Confirmed
Importance: Wishlist
Launchpad Issue: [lp1411476](https://bugs.launchpad.net/bugs/1411476)
Tags: library

---

It would be great if Mixxx colour coded the key information, according to the Camelot Wheel/circle of fifths (http://www.mixedinkey.com/Book/How-to-Use-Harmonic-Mixing-2), or some similar notation, where similar colours indicate that keys are harmonious.

### PR DESCRIPTION
**Pull request by danferns**
Hi! I just got the core of this feature working and I'm seeking feedback.

A small rectangle is drawn next to the Key label in the Library. The color of the rectangle is based on the Key of the track.

image

---

I'm concerned about the method used for finding the track's key. Right now, I'm using `KeyUtils::guessKeyFromText` on the Key string, but I don't think that'll work if the user sets custom Key Notation. 

I want to use the value in [`ColumnCache::COLUMN_LIBRARYTABLE_KEY_ID`](https://github.com/danferns/mixxx/blob/7931a0756cd8fb00422806d685c7029530864f14/src/library/basetracktablemodel.cpp#L888). But I haven't been able to figure out how to access this column from the `QModelIndex& index` that is passed into `KeyDelegate::paintItem`. I'm not sure if the column is even available.

Is there a place where I can see how the index is structured?

---

Once complete, this will resolve #7810.

### CODE REVIEW

File: src/library/tabledelegates/keydelegate.cpp
Comment (Swiftb0y) [inline src/library/tabledelegates/keydelegate.cpp:3]: weird include, remove?
```suggestion
```

File: src/library/tabledelegates/keydelegate.h
Comment (Swiftb0y) [inline src/library/tabledelegates/keydelegate.h:4]: also `#include <QObject>`.

**Comment by Swiftb0y** [comment]
> Once complete, this will resolve https://github.com/mixxxdj/mixxx/issues/7810.

I don't think so. That issue requests to color the key depending on how compatible it is to the currently playing track.

File: src/library/tabledelegates/keydelegate.cpp
Comment (Swiftb0y) [inline src/library/tabledelegates/keydelegate.cpp:16]: move into header please.

File: src/library/tabledelegates/keydelegate.cpp
Comment (Swiftb0y) [inline src/library/tabledelegates/keydelegate.cpp:31]: check bounds as well.
```suggestion
        const auto& palette = mixxx::PredefinedColorPalettes::kDefaultKeyColorPalette;
        DEBUG_ASSERT(openKeyNumber <= palette.length());
        const auto rgbColor =
                palette.at(
                        openKeyNumber - 1); // Open Key numbers start from 1
```

File: src/util/color/predefinedcolorpalettes.cpp
Comment (Swiftb0y) [inline src/util/color/predefinedcolorpalettes.cpp:161]: as a follow up, it would be nice if the palette was customizable via the `Color` Preferences. similar to track colors.

**Comment by Swiftb0y** [comment]
> I'm concerned about the method used for finding the track's key. Right now, I'm using KeyUtils::guessKeyFromText on the Key string, but I don't think that'll work if the user sets custom Key Notation.

Probably not. You should store the underlying `mixxx::track::io::key::ChromaticKey` in the cell when its text gets set. You can use a custom [item role](https://doc.qt.io/qt-6/model-view-programming.html#item-roles) for that (I think, I'm not much into Qt MV programming tbh).

**Comment by danferns** [comment]
Thank you so much @Swiftb0y for the feedback! Preferences for key color palettes is definitely in the plans. I will apply the requested changes and try the item role method.

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [inline src/library/tabledelegates/keydelegate.cpp:46]:
```suggestion
            keyText,
```

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [inline src/library/tabledelegates/keydelegate.cpp:32]: Just a note: we could move the QString -> QColor translation to KeyUtils so it's available in WKey, too (or some other widget).

File: src/util/color/predefinedcolorpalettes.cpp
Comment (ronso0) [reply src/util/color/predefinedcolorpalettes.cpp:161]: Super nice to have, follow up²: set the colors in the notation table in Key preferences

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [inline src/library/tabledelegates/keydelegate.cpp:49]: Like in LocationDelegate we need to get the highlight color from the palette in order to paint the selected key text correctly:
https://github.com/mixxxdj/mixxx/blob/a48562ad96021bd9571c8fbc8d331291bc30c1fc/src/library/tabledelegates/locationdelegate.cpp#L11-L23

**Review by ronso0 — DISMISSED** [review]
Thank you, this looks nice!

File: src/library/basetracktablemodel.cpp
Comment (daschuer) [inline src/library/basetracktablemodel.cpp:125]: No real issue here in this case. In general we try to avoid reformating untouched code to keep the change set minimal. Make sure your IDE does not automatically reformat whole files.

File: src/library/tabledelegates/keydelegate.cpp
Comment (daschuer) [reply src/library/tabledelegates/keydelegate.cpp:16]: What's the reason for this request?

File: src/library/tabledelegates/keydelegate.cpp
Comment (daschuer) [inline src/library/tabledelegates/keydelegate.cpp:24]: Here you should request the color directly from the model using a role, the get around the guessing below.

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/library/tabledelegates/keydelegate.cpp
Comment (Swiftb0y) [reply src/library/tabledelegates/keydelegate.cpp:16]: potentially performance. constructors this trivial will likely always get inlined, but the compiler needs it in the header for that.

File: src/library/basetracktablemodel.cpp
Comment (danferns) [reply src/library/basetracktablemodel.cpp:125]: Thank you for noting that, I've now set it to only format changes.

File: src/library/tabledelegates/keydelegate.cpp
Comment (danferns) [reply src/library/tabledelegates/keydelegate.cpp:36]: Thank you! Just fixed this.

File: src/track/keyutils.cpp
Comment (daschuer) [inline src/track/keyutils.cpp:477]:
```suggestion
        return {}; // return invalid color
```

File: src/track/keyutils.cpp
Comment (daschuer) [inline src/track/keyutils.cpp:473]:
```suggestion
        DEBUG_ASSERT(openKeyNumber <= palette.size() && openKeyNumber >= 1);
```

File: src/util/color/predefinedcolorpalettes.cpp
Comment (Swiftb0y) [reply src/util/color/predefinedcolorpalettes.cpp:161]: Yeah, Though I guess we'll have to decide on whether its better to put in the Color or the Key settings. The former would be more consistent IMO.

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [inline src/library/tabledelegates/keydelegate.cpp:23]: I suggest to remove the left padding in order to allow narrow column configurations.
Would look like this
image

File: src/library/tabledelegates/keydelegate.cpp
Comment (Swiftb0y) [reply src/library/tabledelegates/keydelegate.cpp:16]: yes, guessing the key from the text is inefficient and slow.

File: src/library/tabledelegates/keydelegate.cpp
Comment (danferns) [reply src/library/tabledelegates/keydelegate.cpp]: Done. Thank you for this suggestion!

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [reply src/library/tabledelegates/keydelegate.cpp:16]: Ah, okay. So just get the raw value (key code) via index->data(), cast to int and get the color from `KeyUtils::colorFromNumericValue(int)`?

Comment (ronso0) [reply src/library/tabledelegates/keydelegate.cpp:16]: = don't get the **color** from the model, as requested, just the value

Comment (Swiftb0y) [reply src/library/tabledelegates/keydelegate.cpp:16]: yeah, though I wouldn't even go over the `int` index and just store and retrieve the `mixxx::track::io::key::ChromaticKey` directly. otherwise we'll have to deal with out-of-bound values and so forth again.

File: src/util/color/predefinedcolorpalettes.cpp
Comment (danferns) [reply src/util/color/predefinedcolorpalettes.cpp:161]: Here's a little sketch:

image

Comment (Swiftb0y) [reply src/util/color/predefinedcolorpalettes.cpp:161]: I'd somehow try to put it in the same column. but I'm not a UI designer. Wdyt @ronso0?

Comment (danferns) [reply src/util/color/predefinedcolorpalettes.cpp:161]: I think it looks more compact:

image

Comment (Swiftb0y) [reply src/util/color/predefinedcolorpalettes.cpp:161]: maybe swap the order of the text and the check box so its all still right aligned? idk, IIANA Designer.

**Comment by daschuer** [comment]
Can you update the screenshot in the initial description of this PR? This way it is more significant if one is later following links from the Changelog.

**Comment by daschuer** [comment]
Is the draft state still correct or is it ready for final review and tests?

File: src/util/color/predefinedcolorpalettes.cpp
Comment (daschuer) [reply src/util/color/predefinedcolorpalettes.cpp:147]: https://www-datylon-com.cdn.ampproject.org/v/s/www.datylon.com/blog/data-visualization-for-colorblind-readers?amp_gsa=1&amp_js_v=a9&hs_amp=true&usqp=mq331AQIUAKwASCAAgM%3D#amp_tf=Von%20%251%24s&aoh=17198172132733&csi=1&referrer=https%3A%2F%2Fwww.google.com&ampshare=https%3A%2F%2Fwww.datylon.com%2Fblog%2Fdata-visualization-for-colorblind-readers

Interesting are the hue curves of different color blind people. So I guess in a follow up PR, we may offer the full hue curve and reduced versions for color blind people that is only using half of the hue range.

**Review by daschuer — APPROVED** [review]
Just a minor comment. Rest LGTM, thank you.

**Comment by danferns** [comment]
> Is the draft state still correct or is it ready for final review and tests?

Yes, I am not done yet. I still have to read the Key Colors setting and only render the rectangle if it is enabled. 

Also I'd like to squash a couple of commits before review, at least the one which was failing to build.

**Comment by danferns** [comment]
I want to read the setting using `ConfigKey("[Config]", "KeyColorsEnabled")` inside `KeyUtils::keyToColor` (if key colors are disabled, return an invalid color). But to do that, I would need access to the [`pConfig`](https://github.com/mixxxdj/mixxx/wiki/developer_guide_config#getting-and-setting-preference-options) variable inside keyutils.cpp. 

I see that the `pConfig` is passed around in class constructors. Is there a way to access it directly?

Should I go about this in some other way entirely?

**Comment by Swiftb0y** [comment]
I don't think accessing the config in the KeyUtils is a good idea. Can you share a little more insight on what you're trying to accomplish? Then we can brainstorm together.

**Comment by Swiftb0y** [comment]
Thanks for the explanation.
I think that check should stay inside `KeyDelegate`. The static data in `KeyUtils` is ugly and its usefulness is questionable IMO.

File: src/util/color/predefinedcolorpalettes.cpp
Comment (Swiftb0y) [reply src/util/color/predefinedcolorpalettes.cpp:147]: We could provide preset palettes optimized for the different types of colorblindness. I think gradients for those already exist online. This would be fairly straightforward to implement once the requusted custom keypalettes are in place.

**Comment by danferns** [comment]
Got it, thank you!

**Comment by daschuer** [comment]
In other cases like that we have added a ControlObject 
With the same name. Than you can connect to the value change signal inside the library view and refresh the whole view. I think this is later also required when changing the color schema. So you may consider to add a "None" shema for disable colors.

**Comment by daschuer** [comment]
Like here:
https://github.com/mixxxdj/mixxx/blob/7b018fbe588f9c43270e63393179f5cc5cce9e6c/src/preferences/dialog/dlgprefkey.cpp#L48

**Comment by ronso0** [comment]
> I was planning on calling `slotRefreshAllRows()` in the new [`BaseTrackTableModel::setKeyColorsEnabled`](https://github.com/mixxxdj/mixxx/blob/c0657253e34f141f17a37740ef20e4e8b2790b65/src/library/basetracktablemodel.cpp#L122) method, but then realized that this method needs to be static, while `slotRefreshAllRows` is not static.

I'd recommend to avoid that method. That refreshes the entire view and resorts, i.e. like purging tracks this has the undesired resorting side effect, see #12565 

I see that both the new color state and s_bpmColumnPrecision are applied in the next paint event, which happens for me when I close the preferences and see the table being redrawn. Don't know if that happens when the main window is raised again or only if the preferences covered the tracks table.

**Comment by ronso0** [comment]
Btw the key color default is ON, but the checkbox is not checked when opening the Color pref page.

File: src/preferences/dialog/dlgprefcolors.cpp
Comment (ronso0) [inline src/preferences/dialog/dlgprefcolors.cpp:87]: slotUpdate() is supposed to read the config values and update the view accordingly.
Here, the previous (GUI) value is used (that's why the box is still unchecked even though the setting is ON), and it's not obvious that `m_colorPaletteSettings.get..` queries the config.
Please read the config with `m_pConfig->getValue([color setting])`.

**Review by ronso0 — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
FWIW I expected the key color setting in the Library preferences since that is what handle the settings and appearance of the tracks table.

File: src/preferences/dialog/dlgprefcolors.cpp
Comment (ronso0) [inline src/preferences/dialog/dlgprefcolors.cpp:365]: This slot is called when the checkbox is clicked (changed), so the checkbox state and the config state are now identical.
No need to update the entire page now.

Comment (ronso0) [reply src/preferences/dialog/dlgprefcolors.cpp:365]: Sorry for the fragmented review, I didn't take a closer look earlier.

File: src/preferences/dialog/dlgprefcolorsdlg.ui
Comment (danferns) [reply src/preferences/dialog/dlgprefcolorsdlg.ui]: Fixed, thank you.

File: src/preferences/dialog/dlgprefcolors.cpp
Comment (danferns) [reply src/preferences/dialog/dlgprefcolors.cpp:365]: I did not quite get you. Did you mean that we don't need the `slotUpdate()` call because it gets called some other way when the checkbox is clicked?

Comment (ronso0) [reply src/preferences/dialog/dlgprefcolors.cpp:365]: Yes, we don't need it because the checkbox state and the config state are identical after writing to the config and `setKeyColorsEnabled()`.

**Comment by ronso0** [comment]
> FWIW I expected the key color setting in the Library preferences since that is what handle the settings and appearance of the tracks table.

What do you think?
Should we have the enable switch in the Library page, and the future palette config in Colors?

**Comment by danferns** [comment]
> What do you think? Should we have the enable switch in the Library page, and the future palette config in Colors?

If we add key color support to the WKey widget in the future, then it feels more natural to me to have this setting under the Color preferences. Because then this toggle is not only changing the library view but also the deck view.

**Comment by danferns** [comment]
Thank you @daschuer, @ronso0 and @Swiftb0y for your feedback so far! I think I'm ready for a final review.

I was thinking of cleaning up the git log a little, but I think I'll do it at the end before merging (would that be fine?). There might still be more changes to be made from now, and I'm also not sure how much to clean up.

**Comment by Swiftb0y** [comment]
Yes, don't worry with cleaning up the log for now. I'm sure there will be a bunch more comments.

**Comment by danferns** [comment]
I want to work on adding the key color palettes. I'm a bit unsure which branch to do it on. Should I branch off from this PR's branch, and then create a separate PR for the palettes? Or would it be better to keep working on this branch only?

**Comment by ronso0** [comment]
I suggest to make "small" steps, i.e. get this merged first, then improve / tweak / extend this new feature in separate branches.

**Comment by danferns** [comment]
Got it, thank you.

**Review by ronso0 — COMMENTED** [review]
Alrighty, this LGTM except a few bits.

**Comment by ronso0** [comment]
> maybe swap the order of the text and the check box so its all still right aligned? idk, IIANA Designer.

IMHO it's okay like this.
IIRC it's not possible to swap checkbox and label, that would need to be two separate widgets (QLabel + QCheckBox without label).

File: src/library/tabledelegates/keydelegate.cpp
Comment (ronso0) [inline src/library/tabledelegates/keydelegate.cpp:36]: not the other way around, since we already have a QString and don't need to construct another one?
(based what I think I understood fo QString stuff)
```suggestion
    if (keyText.compare("INVALID") == 0) { // if the key is invalid, display empty string
```

Comment (Swiftb0y) [reply src/library/tabledelegates/keydelegate.cpp:36]: good suggestion, but better to avoid string comparison altogether. https://github.com/mixxxdj/mixxx/pull/13390#discussion_r1672347393

**Comment by danferns** [comment]
Thank you for the review. I have used the enum comparison now, and placed all the text/color rendering code inside this condition. 

Let me know when it is the right time for rebasing.

**Review by ronso0 — APPROVED** [review]
Yep, LGTM : )

File: src/preferences/colorpalettesettings.h
Comment (Swiftb0y) [inline src/preferences/colorpalettesettings.h:6]: can you remove the `#define`? just copy paste it into its only use below.

**Review by Swiftb0y — COMMENTED** [review]
not perfect, but good enough considering the status quo. almost LGTM.

**Comment by danferns** [comment]
> not perfect, but good enough considering the status quo.

I appreciate the honest feedback. Can you please tell me in what aspects I can improve?

**Comment by Swiftb0y** [comment]
> I appreciate the honest feedback. 

Don't get me wrong, this PR is quite great already ;). There is just always room to improve. 

> Can you please tell me in what aspects I can improve?

The primary issue I have is that you followed some suboptimal patterns that are already in the existing code. I'm not expecting you to clean those up, though I'm not particularly fond of the fact that this PR is adding more suboptimal code. Refactoring the code in question is probably outside your skillset and GSoC project scope but there are a bunch of refactorings that are in order IMO. Ideally, all the static global variables (the ones prefixed `s_`) should be eliminated (not sure how, haven't looked too deep into it). `src/util/color/predefinedcolorpalettes.cpp` could use a little bit of a cleanup. `KeyDelegate::paintItem` could use a little more reordering to simplify its controlflow (prefer early returns) and separate the individual steps better (currently a lot of the low-level math of the drawing code is interspersed with the high-level drawing code) (see [SLAP](http://principles-wiki.net/principles:single_level_of_abstraction)). Passing `s_keyColorsEnabled` as part of the `BaseTrackTableModel` is weird (especially via the `roleValue()`) because the value doesn't change depending on the index.

Having said all that, I'm not expecting you to address these concerns. Instead discuss with your mentor (@daschuer IIRC?) on how to proceed.

**Review by Swiftb0y — APPROVED** [review]
LGTM. Thank you. Who wants to press the button? Did you want to rebase first?

**Comment by danferns** [comment]
Thank you so much, this is very helpful. I'll discuss with Daniel for the next steps, and I'll keep these points in mind for future PRs. I am growing a lot thanks to this project. :) 

I do want to rebase!

**Comment by daschuer** [comment]
Thank you for the review. Yes, there is room for improvements. On the other hand we have tight plan for the remaining project. From this point of view I would rather see a beautiful Heat column instead of beautiful code. 

We have already the green check marks so I will merge once @danferns gives a go. 

If you feeling lucky to address some of @Swiftb0y points, go ahead and propose it in a separate PR.

**Comment by danferns** [comment]
I'm done with the rebase! I squashed only a build error, and two commits involving indentation / comments. If anything else is needed then let me know, but otherwise I am ready for this PR to be merged!

**Comment by Swiftb0y** [comment]
> On the other hand we have tight plan for the remaining project. From this point of view I would rather see a beautiful Heat column instead of beautiful code.

I agree in this case.

**Comment by daschuer** [comment]
Thank you very much! This looks really nice and works good.

**Comment by acolombier** [comment]
After synching my dev branch, I just seen that eye candy appeared. It looks very good! Thanks for that feature!

**Comment by Eve00000** [comment]
I tested this today, some remarks:
- I don't have a key palette under 'Enable Key Colors' in Preferences -> Colors
-Pushing 'Apply' after changing 'Enable Key Colors' does not work, changes only happen after push on 'OK'
- Nice color bar in 'Key'-column but Key text is cut of on the left side (normally text is cut of on the right side) while my most important part is on the left side.
afbeelding
afbeelding

**Comment by ronso0** [comment]
> * I don't have a key palette under 'Enable Key Colors' in Preferences -> Colors

There is only one palette atm.

Repaint with/without colors only happens when the table is repainted after the preferences are closed.

**Pull request merged**
